README.md
# twilio-go

A client for accessing the Twilio API with several nice features:

- Easy-to-use helpers for purchasing phone numbers and sending MMS messages

- E.164 support, times that are parsed into a time.Time, and other smart types.

- Finer grained control over timeouts with a Context, and the library uses
  wall-clock HTTP timeouts, not socket timeouts.

- Easy debugging network traffic by setting DEBUG_HTTP_TRAFFIC=true in your
  environment.

- Easily find calls and messages that occurred between a particular
set of `time.Time`s, down to the nanosecond, with GetCallsInRange /
GetMessagesInRange.

- It's clear when the library will make a network request, there are no
unexpected latency spikes when paging from one resource to the next.

- Uses threads to fetch resources concurrently; for example, has methods to
fetch all Media for a Message concurrently.

- Usable, [one sentence descriptions of Alerts][alert-descriptions].

[alert-descriptions]: https://godoc.org/github.com/mosessky/twilio-go#Alert.Description

Here are some example use cases:

```go
const sid = "AC123"
const token = "456bef"

client := twilio.NewClient(sid, token, nil)

// Send a message
msg, err := client.Messages.SendMessage("+14105551234", "+14105556789", "Sent via go :) ✓", nil)

// Start a phone call
var callURL, _ = url.Parse("https://kev.inburke.com/zombo/zombocom.mp3")
call, err := client.Calls.MakeCall("+14105551234", "+14105556789", callURL)

// Buy a number
number, err := client.IncomingNumbers.BuyNumber("+14105551234")

// Get all calls from a number
data := url.Values{}
data.Set("From", "+14105551234")
callPage, err := client.Calls.GetPage(context.TODO(), data)

// Iterate over calls
iterator := client.Calls.GetPageIterator(url.Values{})
for {
    page, err := iterator.Next(context.TODO())
    if err == twilio.NoMoreResults {
        break
    }
    fmt.Println("start", page.Start)
}
```

A [complete documentation reference can be found at
godoc.org](https://godoc.org/github.com/mosessky/twilio-go).

## In Production

twilio-go is being used by the following applications:

- [Logrole][logrole], an open source Twilio log viewer that's faster than the API.

Using twilio-go in production? [Let me know!](mailto:kevin@burke.services)

[logrole]: https://github.com/kevinburke/logrole

## Supported API's

The API is unlikely to change, and currently covers these resources:

- Alerts
- Applications
- Calls
- Conferences
- Faxes
- Incoming Phone Numbers
- Available Phone Numbers
- Keys
- Messages
- Media
- Monitor
- Outgoing Caller ID's
- Pricing
- Queues
- Recordings
- Task Router
  - Activities
  - TaskQueues
  - Workers
  - Workflows
- Transcriptions
- Wireless
- Access Tokens for IPMessaging, Video and Programmable Voice SDK

### Error Parsing

If the twilio-go client gets an error from the Twilio API, we attempt to convert
it to a [`rest.Error`][rest.error] before returning. Here's an example 404.

[rest.error]: https://godoc.org/github.com/kevinburke/rest#Error

```
&rest.Error{
    Title: "The requested resource ... was not found",
    ID: "20404",
    Detail: "",
    Instance: "",
    Type: "https://www.twilio.com/docs/errors/20404",
    StatusCode: 404
}
```

Not all errors will be a `rest.Error` however - HTTP timeouts, canceled
context.Contexts, and JSON parse errors (HTML error pages, bad gateway
responses from proxies) may also be returned as plain Go errors.

### Twiml Generation

There are no plans to support Twiml generation in this library. It may be
more readable and maintainable to manually write the XML involved in a Twiml
response.

### Errata

- Media URL's used to be returned over HTTP. twilio-go rewrites these URL's to be
  HTTPS before returning them to you.

- A subset of Notifications returned code 4107, which doesn't exist. These
  notifications should have error code 14107. We rewrite the error code
  internally before returning it to you.

- The only provided API for filtering calls or messages by date grabs all
messages for an entire day, and the day ranges are only available for UTC. Use
GetCallsInRange or GetMessagesInRange to do timezone-aware, finer-grained date
filtering.

- You can get Alerts for a given Call or MMS by passing `ResourceSid=CA123` as
a filter to Alerts.GetPage. This functionality is not documented in the API.


CHANGELOG.md
# Changes

## 2.1

- Support deleting messages via `client.Messages.Delete(sid)`.

- Support a different default region for local phone number parsing than "US" by
  overriding the value of `twilio.DefaultRegion` in an init function.

Thanks Kevin Golding for the suggestions and issue reports.

## 2.0

Update Pricing API to v2. This is a breaking change.

To upgrade from the v1 Pricing API:

* `VoicePrice` is now `VoicePrices`.
* `VoiceNumberPrice` is now `VoiceNumberPrices`.
* `NumberPrice` is now `NumberPrices`.
* `MessagePrice` is now `MessagePrices`
* The `Prefixes` field in the `PrefixPrice` struct is now `DestinationPrefixes`.
* `OriginationPrefixes` field has been added to the `PrefixPrice` struct.
* The `OutboundCallPrice` field in the `VoiceNumberPrices` struct is now
`OutboundCallPrices`. Add the old `OutboundCallPrice` as the first entry in the
array.
* `OriginationPrefixes` field has been added to the `OutboundCallPrice` struct.

* `NumberVoicePriceService.Get` now has 2 parameters: `destinationNumber` and
`data` url.Values. `destinationNumber` is same as the previous `number` field.
`data` can contain additional parameters like `OriginationNumber`.
* Several other `*PriceService.Get` functions now have a third `url.Values`
  parameter. Pass `nil` for this third parameter to maintain compatibility with
  the v1 API.

## 1.8

Add Task Router Workers API.

## 1.7

Add Task Router Workflow API.

## 1.6

Add Task Router TaskQueue API.

## 1.5

Add Task Router Activity API.

## 1.4

Add Supported Countries service. (via Andrei Schneider @megaherz)

## 1.3

Remove dependency on jwt-go (functionality is the same, we just generate JWT's
by hand now.)

## 1.2

Fix key for ConferenceSid.

## 1.1

Add video and video recordings API

## 1.0

Update kevinburke/rest to version 2.0.

## 0.66

Support the Notify Credentials API. Thanks to Maksym Pavlenko for implementing
this change.

## 0.65

Support the Commands Wireless API.

Add cmd/report-data-usage and cmd/report-data-usage-server for reporting recent
data usage for Sims.

## 0.64

Partial support for wireless.twilio.com endpoints, including the Sim resource
and UsageRecords for Sims.

This version of the library requires github.com/kevinburke/go-types version
0.20 at least.

## 0.63

Support Update() for the IncomingPhoneNumber resource.

## 0.62

Fix error in release 0.61.

## 0.61

Support the Available Phone Numbers API. Thanks to Maksym Pavlenko for
contributing the patch.

## 0.59

Support the Twilio Fax API.

- Rename imports from github.com/saintpete/twilio-go to
  github.com/mosessky/twilio-go.

## 0.58

Add `client.UseSecretKey(key string)` to handle secret keys
properly. For more information on the secret key API, see
https://www.twilio.com/docs/api/rest/keys.

Thanks to Andrew Watson for reporting.

## 0.57

Switch all imports to use the `"context"` library.

## 0.56

Use the new github.com/kevinburke/rest.DefaultTransport RoundTripper for
easy HTTP debugging. (The previous code set the RoundTripper to nil, so
kevinburke/rest wouldn't log anything).

## 0.55

Handle new HTTPS-friendly media URLs.

Support deleting/releasing phone numbers via IncomingNumbers.Release(ctx, sid).

Initial support for the Pricing API (https://pricing.twilio.com).

Add an AUTHORS.txt.

## 0.54

Add Recordings.GetTranscriptions() to get Transcriptions for a recording. The
Transcriptions resource doesn't support filtering by Recording Sid.

## 0.53

Add Alert.StatusCode() function for retrieving a HTTP status code (if one
exists) from an alert.

## 0.52

Copy url.Values for GetXInRange() functions before modifying them.

## 0.51

Implement GetNextConferencesInRange

## 0.50

Implement GetConferencesInRange. Fix paging error in
GetCallsInRange/GetMessagesInRange.

## 0.47

Implement GetNextXInRange - if you have a next page URI and want to get an
Iterator (instead of starting with a url.Values).

## 0.45

Fix go 1.6 (messages_example_test) relied on the stdlib Context package by
accident.

## 0.44

Support filtering Calls/Messages down to the nanosecond in a TZ-aware way, with
Calls.GetCallsInRange / Messages/GetMessagesInRange.

## 0.42

Add more Description fields based on errors I've received in the past. There
are probably more to be found, but this is a good start.

## 0.41

Use the same JWT library instead of using two different ones.

Add Description() for Alert bodies.

## 0.40

Fix next page URL's for Twilio Monitor

## 0.39

The data in Update() requests was silently being ignored. They are not ignored
any more.

Support the Accounts resource.

Add RequestOnBehalfOf function to make requests on behalf of a subaccount.

Fixes short tests that were broken in 0.38

## 0.37

Support Outgoing Caller ID's

## 0.36

Support Keys

## 0.35

Added Ended(), EndedUnsuccessfully() helpers to a Call, and FriendlyPrice() to
a Transcription.


example_test.go
package twilio_test

import (
	"context"
	"fmt"
	"log"
	"net/url"
	"time"

	"github.com/kevinburke/rest"
	twilio "github.com/mosessky/twilio-go"
)

var callURL, _ = url.Parse("https://kev.inburke.com/zombo/zombocom.mp3")
var pdfURL, _ = url.Parse("https://kev.inburke.com/foo.pdf")

func Example() {
	client := twilio.NewClient("AC123", "123", nil)

	// Send a SMS
	msg, _ := client.Messages.SendMessage("+14105551234", "+14105556789", "Sent via go :) ✓", nil)
	fmt.Println(msg.Sid, msg.FriendlyPrice())

	// Make a call
	call, _ := client.Calls.MakeCall("+14105551234", "+14105556789", callURL)
	fmt.Println(call.Sid, call.FriendlyPrice())

	_, err := client.IncomingNumbers.BuyNumber("+1badnumber")
	// Twilio API errors are converted to rest.Error types
	if err != nil {
		restErr, ok := err.(*rest.Error)
		if ok {
			fmt.Println(restErr.Title)
			fmt.Println(restErr.Type)
		}
	}

	// Find all calls from a number
	data := url.Values{"From": []string{"+14105551234"}}
	iterator := client.Calls.GetPageIterator(data)
	ctx, cancel := context.WithTimeout(context.TODO(), 1*time.Minute)
	defer cancel()
	for {
		page, err := iterator.Next(ctx)
		if err == twilio.NoMoreResults {
			break
		}
		for _, call := range page.Calls {
			fmt.Println(call.Sid, call.To)
		}
	}
}

func ExampleCallService_GetCallsInRange() {
	// Get all calls between 10:34:00 Oct 26 and 19:25:59 Oct 27, NYC time.
	nyc, _ := time.LoadLocation("America/New_York")
	start := time.Date(2016, 10, 26, 22, 34, 00, 00, nyc)
	end := time.Date(2016, 10, 27, 19, 25, 59, 00, nyc)

	client := twilio.NewClient("AC123", "123", nil)
	iter := client.Calls.GetCallsInRange(start, end, url.Values{})
	ctx, cancel := context.WithTimeout(context.TODO(), 10*time.Second)
	defer cancel()
	for {
		page, err := iter.Next(ctx)
		if err == twilio.NoMoreResults {
			break
		}
		if err != nil {
			log.Fatal(err)
		}
		for i, call := range page.Calls {
			fmt.Printf("%d: %s (%s)", i, call.Sid, call.DateCreated.Time)
		}
	}
}

func ExampleClient_UseSecretKey() {
	client := twilio.NewClient("AC123", "123", nil)
	client.UseSecretKey("SK123")
	client.Messages.SendMessage("123", "456", "Sending with secret key...", nil)
}

func ExampleFaxService_SendFax() {
	faxer := twilio.NewFaxClient("AC123", "123", nil)
	faxer.Faxes.SendFax("123", "456", pdfURL)
}

func ExampleFaxService_Cancel() {
	faxer := twilio.NewFaxClient("AC123", "123", nil)
	faxer.Faxes.Cancel("FX123")
}

func ExampleFax() {
	faxer := twilio.NewFaxClient("AC123", "123", nil)
	fax, _ := faxer.Faxes.Get(context.TODO(), "FX123")
	fmt.Print(fax.Sid)
}

func ExampleNewTaskRouterClient() {
	client := twilio.NewTaskRouterClient("AC123", "123", nil)
	data := url.Values{"FriendlyName": []string{"On Call"}}
	client.Workspace("WS123").Activities.Create(context.TODO(), data)
}


token/example_test.go
package token_test

import (
	"fmt"
	"time"

	"github.com/mosessky/twilio-go/token"
)

func Example() {
	t := token.New("AC123", "456bef", "secretkey", "test@example.com", time.Hour)
	grant := token.NewConversationsGrant("a-conversation-sid")
	t.AddGrant(grant)
	jwt, _ := t.JWT()
	fmt.Println(jwt) // A string encoded with the given values.
}


alerts_example_test.go
package twilio_test

import (
	"context"
	"fmt"
	"log"
	"net/url"

	twilio "github.com/mosessky/twilio-go"
)

func ExampleAlertService_GetPage() {
	client := twilio.NewClient("AC123", "123", nil)
	data := url.Values{}
	data.Set("ResourceSid", "SM123")
	page, err := client.Monitor.Alerts.GetPage(context.TODO(), data)
	if err != nil {
		log.Fatal(err)
	}
	for _, alert := range page.Alerts {
		fmt.Println(alert.Sid)
	}
}


datausage/datausage.go
// Package datausage has utilities for retrieving data usage from a SIM card.
package datausage

import (
	"context"
	"errors"
	"net/url"
	"time"

	"golang.org/x/sync/errgroup"

	types "github.com/kevinburke/go-types"
	twilio "github.com/mosessky/twilio-go"
)

// GetUsage gets usage for the given sim, starting at start and ending at end.
// The return value is a list of usage information. The first item in the list
// is the interval starting at start and so on until end is reached.
//
// sim may be either a Sim SID or a UniqueName.
func GetUsage(ctx context.Context, client *twilio.Client, sim string, start, end time.Time, interval time.Duration) ([]types.Bits, error) {
	if interval <= 0 {
		panic("invalid interval: " + interval.String())
	}
	iter := start
	count := 0
	for {
		count++
		iter = iter.Add(interval)
		if !iter.Before(end) {
			break
		}
	}
	list := make([]types.Bits, count)
	group, errctx := errgroup.WithContext(ctx)
	for i := 0; i < count; i++ {
		i := i
		group.Go(func() error {
			intervalStart := start.Add(interval * time.Duration(i)).Format(time.RFC3339)
			intervalEnd := start.Add(interval * time.Duration(i+1)).Format(time.RFC3339)
			page, err := client.Wireless.Sims.GetUsageRecords(errctx, sim, url.Values{
				"Start": []string{intervalStart},
				"End":   []string{intervalEnd},
			})
			if err != nil {
				return err
			}
			if len(page.UsageRecords) == 0 {
				return errors.New("no usage records for date range")
			}
			if len(page.UsageRecords) > 1 {
				return errors.New("too many usage records for date range")
			}
			list[i] = page.UsageRecords[0].Data.Total
			return nil
		})
	}
	if err := group.Wait(); err != nil {
		return nil, err
	}
	return list, nil
}


cmd/report-data-usage/main.go
// Binary report-data-usage reports information about how much data your sim cards have used.
//
// Example report format:
//
//     $ report-data-usage
//     iPhone v13
//     ----------
//     2017-06-19: 0
//     2017-06-20: 0
//     2017-06-21: 11.281MB
//     2017-06-22: 93.341MB
//     2017-06-23: 94.422MB
//     2017-06-24: 159.461MB
//     2017-06-25: 50.062MB
//     total (last 7 days): 408.568MB
//
// Your Twilio credentials are loaded from the TWILIO_ACCOUNT_SID and
// TWILIO_AUTH_TOKEN environment variables. There are several flags:
//
//     --days int
//         Change the number of days to report usage for (default 7)
//     --location string
//         Use a different timezone for day boundaries (example "America/Los_Angeles")
//     --sim string
//         Only fetch usage for this sim
package main

import (
	"context"
	"flag"
	"fmt"
	"log"
	"os"
	"strings"
	"sync"
	"time"

	types "github.com/kevinburke/go-types"
	twilio "github.com/mosessky/twilio-go"
	"github.com/mosessky/twilio-go/datausage"
	"golang.org/x/sync/errgroup"
)

var duration = flag.Uint("days", 7, "Number of days to get usage for")
var location = flag.String("location", "", "Timezone to use (defaults to system location/TZ env var)")
var sim = flag.String("sim", "", "Load data about a specific Sim/unique name only")

func init() {
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, `Print information about how much data you used on each day.
`)
		flag.PrintDefaults()
	}
}

func main() {
	flag.Parse()
	var loc *time.Location
	if *location == "" {
		loc = time.Local
	} else {
		var err error
		loc, err = time.LoadLocation(*location)
		if err != nil {
			log.Fatal(err)
		}
	}
	c := twilio.NewClient(os.Getenv("TWILIO_ACCOUNT_SID"), os.Getenv("TWILIO_AUTH_TOKEN"), nil)
	end := time.Now().Add(24 * time.Hour).In(loc)
	end = time.Date(end.Year(), end.Month(), end.Day(), 0, 0, 0, 0, end.Location())
	start := end.Add(-time.Duration(*duration+1) * 24 * time.Hour)
	iter := c.Wireless.Sims.GetPageIterator(nil)
	sims := make([]string, 0)
	if *sim != "" {
		sims = []string{*sim}
	} else {
		for {
			ctx, cancel := context.WithTimeout(context.Background(), 31*time.Second)
			page, err := iter.Next(ctx)
			cancel()
			if err == twilio.NoMoreResults {
				break
			}
			if err != nil {
				log.Fatal(err)
			}
			for i := 0; i < len(page.Sims); i++ {
				sims = append(sims, page.Sims[i].UniqueName)
			}
		}
	}
	// TODO we could probably get fancy here and send the sims on channels when
	// we get them above, which would give us a head start on fetching the data
	//
	// we could also guard data fetches with a semaphore:
	// github.com/kevinburke/semaphore
	var mu sync.Mutex
	ctx, cancel := context.WithTimeout(context.Background(), 31*time.Second)
	defer cancel()
	group, errctx := errgroup.WithContext(ctx)
	usage := make(map[string][]types.Bits)
	for _, name := range sims {
		name := name
		group.Go(func() error {
			u, err := datausage.GetUsage(errctx, c, name, start, end, 24*time.Hour)
			if err != nil {
				return err
			}
			mu.Lock()
			usage[name] = u
			mu.Unlock()
			return nil
		})
	}
	if err := group.Wait(); err != nil {
		log.Fatal(err)
	}
	for _, name := range sims {
		total := types.Bits(0)
		fmt.Printf("%s\n%s\n", name, strings.Repeat("-", len(name)))
		for i := 0; i < len(usage[name]); i++ {
			t := start.Add(time.Duration(i) * 24 * time.Hour)
			fmt.Printf("%s: %s\n", t.Format("2006-01-02"), usage[name][i])
			total += usage[name][i]
		}
		fmt.Printf("total (last %d days): %s\n\n", *duration, total)
	}
}
